oci/spec_opts_windows.go

```diff
@@ -100,3 +101,52 @@ func escapeAndCombineArgs(args []string) string {
 	}
 	return strings.Join(escaped, " ")
 }
+
+func WithDevices(devicePath, containerPath, permissions string) SpecOpts {
+	return func(_ context.Context, _ Client, _ *containers.Container, s *Spec) error {
+		if containerPath != "" {
+			return fmt.Errorf("unexpected containerPath %s", containerPath)
+		}
+
+		if permissions != "" {
+			return fmt.Errorf("unexpected permissions %s", permissions)
+		}
+
+		if devicePath == "" {
+			return errors.New("missing devicePath")
+		}
+
+		// Unambiguously split on ://
+		if splitParts := strings.SplitN(devicePath, "://", 2); len(splitParts) == 2 {
+			if len(splitParts[0]) == 0 {
+				return fmt.Errorf("missing device type in devicePath %v", devicePath)
+			}
+			if s.Windows == nil {
+				s.Windows = &specs.Windows{}
+			}
+			s.Windows.Devices = append(s.Windows.Devices, specs.WindowsDevice{IDType: splitParts[0], ID: splitParts[1]})
+			return nil
+		}
+
+		// Split on / might accidentally match a POSIX relative path, but that would fail later anyway.
+		if splitPaths := strings.SplitN(devicePath, "/", 2); len(splitPaths) == 2 && splitPaths[0] != "" {
+			if s.Windows == nil {
+				s.Windows = &specs.Windows{}
+			}
+			s.Windows.Devices = append(s.Windows.Devices, specs.WindowsDevice{IDType: splitPaths[0], ID: splitPaths[1]})
+			return nil
+		}
+
+		// Absolute path? Probably a POSIX path, for LCOW containers mounting from their host UVM.
+		if devicePath[0] == '/' {
+			if s.Linux == nil {
+				s.Linux = &specs.Linux{}
+			}
+			s.Linux.Devices = append(s.Linux.Devices, specs.LinuxDevice{Path: devicePath})
+			return nil
+		}
```

Comment:
I'm a little doubtful about this behaviour. I wonder if perhaps I should only attempt this if `s.Linux` already exists, or will that introduce a faulty ordering dependency? I'm not assuming `s.Windows` is populated in this function either, and that seems to be the general practice here.

Reply:
I realised I need to go look at how LCOW is communicated over CRI. I assume it's by runtime, since we just had this exact discussion over the puller platform matcher, and I see in [hcsshim](https://github.com/microsoft/hcsshim/blob/master/test/cri-containerd/main_test.go#L49) the use of "runhcs-lcow" `runtimeHandler`, and I can see how this was roughly intended to work by inspecting the hcsshim `Options` of the runtime config in [this commit in the jterry/cri fork](https://github.com/jterry75/cri/commit/d794e3a7e7be811649bdba9ede5e1cf353734c58), but this flow hasn't landed in upstream containerd as far as I can see.

But it does make sense to me that it's the CRI level that knows if we're WCOW, or Hyper-V isolated LCOW or WCOW, and so the OCI layer shouldn't really be trying to second-guess that. That would mean for _this_ PR, `oci.WithLinuxDevices`/`oci.WithLCOWDevices` would not be called except by tests because we don't actually have handlers for this.

I could implement `pkg/cri/opts.WithDevices` to take `config.Runtime` as well, and in there decide if POSIX-pathed devices are LinuxDevices or errors, but that would be the first place that's checking inside this structure in containerd, and I'm not sure if that's the long-term plan; compared with, e.g. getting 'platform' lifted up into the CRI spec for example.

Reply:
I think this is a good example of the last comment. From the perspective of a hypervisor if we did this we would have:

oci.WithWindowsDevices(IDType=A Device, ID=A Path)
oci.WithDevice(hostPathRelativeToTheGuestUVM, containerPath, permissions)

Problem is that `hostPathRelativeToTheGuestUVM` is something only known to the bridge, which usually means it needs to be `fixed up` in the shim. So it seems like its safe to just consider them as `WithWindowsDevices` is always adding a Windows device formed by a Windows path. Because to the Host, its always Windows. And in the guest, we already have a Linux way to describe a path.

Reply:
I'm not sure I understand this comment. Is the suggesting that we _only_ have `oci.WithWindowsDevices(IDType, ID)`, and in effect change the [hcsshim tests](https://github.com/microsoft/hcsshim/blob/643ef46feef41b347d2d8579247b4bc8e6c236a2/test/cri-containerd/createcontainer_test.go) from 
```
			Devices: []*runtime.Device{
				{
					HostPath: "/dev/fuse",
				},
			},
```
to something like
```
			Devices: []*runtime.Device{
				{
					HostPath: "linux-device:///dev/fuse",
				},
			},
```
with `oci.WithWindowsDevices` knowing that `IDType==linux-device` is redirected to the oci `Spec.Linux.Devices` as a path-only `oci.LinuxDevice`, to be turned into a fuller `oci.LinuxDevice` inside the UVM by [this code](https://github.com/microsoft/hcsshim/blob/05093b3d521603ae81d89692dc7b5b4610d2a842/internal/guest/runtime/hcsv2/spec.go#L245-L252), where it can do the devnode lookups?

Reply:
No, I'm sorta saying that there is no good way to accomplish that type of mount at all from the callers perspective. For example, when mounting a host folder we pass a Mount that says "hostPath: A\Folder\Path" "ContainerPath: Another\Path" but what we actually do is this:

Mount To UVM
HostPath: A\Folder\Path
Guest: A\Temp\Path

Mount For Container
HostPath (InGuest): A\Temp\Path
ContainerPath: Another\path

There would really be no way for the caller of containerd to direct such a scheme. It's impossible to know that the GCS (a private protocol between host and guest) would support such a double mount schema. So for devices it really is the same IMO.

What you mean to say is "Map This Device Into That Container" and from the hosts perspective all you have is a host path (a Windows path) and then the rest is sorta up to the shim. That may take the form of the shim saying it supports something like `linux-device` IDTypes or it may do something else. But it's really a custom shim opt at that point. I don't know if we can make it generic

Reply:
Okay, I think I'm with you. So reject `/dev/fuse`, and if someone passes in `linux-device:///dev/fuse`, that'll appear in Windows Devices, and then the in-UVM shim can recognise that like it currently recognises Linux Devices, and adapt it into a Linux Device internally the same way it does for Linux Devices now.

This means `linux-device` is exactly as defined as `class`, `vpci`, `gpu`, etc, i.e. in the shim implementation.

And if another implementation of Windows-containerd-runs-Linux-containers appears, e.g. "modify and forward to a Linux-containerd instance" (ala the Docker Desktop support for Linux containers), then that "modify" would also need to recognise IDType `linux-device` and turn it into an appropriate Linux Device, because at that layer, it would also need to be rejecting _other_ IDTypes anyway, since they are Windows-specific and shouldn't be given to the Linux-hosted containerd.

That's certainly _easy_ for me to implement here.

I do have a concern about rejecting `/dev/fuse` though: At some point in the future, once we have a distinct CRI type for Windows Devices, it would be very reasonable to expect cri.Device `/dev/fuse` to work, i.e. to (in-effect, but not necessary in-implementation) add `linux-device://` when Windows containerd receives such a cri.Device object.

One option is to reject cri.Device that doesn't contain an ID Type (i.e. starts with `/` or `://`), because future expansion of supported inputs is fairly easy.

Another option is to fallback (as I do now) absolute-pathed cri.Devices to oci.Linux.Devices, because while it's shim-specific what that _means_ when oci.Windows is also set, technically so is `class://`.

Perhaps a third option is to actually allow it through, although the naïve parse if `/dev/fuse` is IDType `""`, ID `dev/fuse`, which _looks_ like a mistake.

Reply:
On the note on the naïve parse, I had also been considering making the change that `://` is the formally-supported separator, and _explicitly_ recognising `class/` prefix as Docker-legacy and rewriting it to be `class://` before continuing parsing.

That would simplify the logic and discussion, and removes any ambiguity around `/dev/fuse`, e.g. type `dev/fuse` would no longer get passed down to hcsshim as IDType `dev`.

Reply:
Resolving as I've rewritten this per https://github.com/containerd/containerd/pull/6618#issuecomment-1060203801

Reply:
For reference, here's how we handle setting up oci devices in our CRI fork https://github.com/kevpar/cri/blob/3aa114c67c0ff4d6ac754ed0f656d5ce13562e1b/pkg/server/container_create_windows.go#L619